test/test_generators.py

```diff
@@ -0,0 +1,122 @@
+from typing import AsyncIterator, Iterator
+
+import jax.numpy as jnp
+import pytest
+
+from jaxtyping import Array, Float, jaxtyped, Shaped
+
+from .helpers import ParamError
+
+
+try:
+    import torch
+except ImportError:
+    torch = None
+
+
+def test_generators_simple(typecheck):
+    @jaxtyped(typechecker=typecheck)
+    def gen(x: Float[Array, "*"]) -> Iterator[Float[Array, "*"]]:
+        yield x
+
+    @jaxtyped(typechecker=typecheck)
+    def foo():
+        next(gen(jnp.zeros(2)))
+        next(gen(jnp.zeros((3, 4))))
+
+    foo()
+
+
+def test_generators_return_no_annotations(typecheck):
+    @jaxtyped(typechecker=None)
+    @typecheck
+    def gen(x: Float[Array, "*"]):
+        yield x
+
+    @jaxtyped(typechecker=None)
+    def foo():
+        next(gen(jnp.zeros(2)))
+        next(gen(jnp.zeros((3, 4))))
+
+    foo()
+
+
+def test_generators_double_decorator(typecheck):
+    @jaxtyped(typechecker=None)
+    @typecheck
+    def gen(x: Float[Array, "*"]) -> Iterator[Float[Array, "*"]]:
+        yield x
+
+    @jaxtyped(typechecker=None)
+    def foo():
+        next(gen(jnp.zeros(2)))
+        next(gen(jnp.zeros((3, 4))))
+
+    foo()
+
+
+@pytest.mark.asyncio
+async def test_async_generators_simple(typecheck):
+    @jaxtyped(typechecker=typecheck)
+    async def gen(x: Float[Array, "*"]) -> AsyncIterator[Float[Array, "*"]]:
+        yield x
+
+    @jaxtyped(typechecker=typecheck)
+    async def foo():
+        async for _ in gen(jnp.zeros(2)):
+            pass
+        async for _ in gen(jnp.zeros((3, 4))):
+            pass
+
+    await foo()
+
+
+@pytest.mark.asyncio
+async def test_async_generators_double_decorator(typecheck):
```

Comment:
Btw, we actually have a helper to handle considering both the old and new syntax:

https://github.com/patrick-kidger/jaxtyping/blob/8de8c0bb68c41dbd7d80a6e373eacae1229efe6a/test/test_array.py#L48

`jaxtyp` and `typecheck` are parameterised test fixtures, so that `jaxtyp(typecheck)` will run both `@jaxtyped(typechcker=typecheck)` and `@jaxtyped(typechecker=None)@typecheck`

Reply:
Nice! Didn't see this originally. Fixed now